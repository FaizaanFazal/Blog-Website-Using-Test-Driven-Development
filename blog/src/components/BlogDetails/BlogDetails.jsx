import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import images from '../../utils/images';
import '../../styles/Card.scss';


export default function BlogDetails() {
  const params = useParams();
  const { slug } = params;
  const blogs = useSelector((state) => state.blog.blogs);
  const blogdetail = blogs.filter((blog) => blog.slug === slug)[0];

  return (
    <section data-testid="details">
      <div className="container">
        <h4 data-testid="titleHeading" className="title title-lg">{blogdetail.title}</h4>
        <div className="card-footer flex flex justify-between items-center">
          <div className="writer-info grid">
            <div className="info-avatar">
              <img src={blogdetail.authorImage.src} alt={blogdetail.authorImage.alt} className="object-fit-cover" />
            </div>
            <div className="info-intro">
              <p className="intro-name text-base">{blogdetail.author}</p>
              {blogdetail.verified
                ? (
                  <div className="intro-verify flex items-center">
                    <img src={`${images.verify_icon_filled}`} alt="icon" className="verify-icon" />
                    <span className="verify-status text-sm text">Verified writer</span>
                  </div>
                )
                : (
                  <div className="intro-verify flex items-center">
                    <img src={`${images.verify_icon_light}`} alt="icon" className="verify-icon" />
                    <span className="verify-status text-sm text">Not-Verified writer</span>
                  </div>
                )}
            </div>
          </div>

          <div className="date-info text text-base" data-testid="detailsdate">{blogdetail.date}</div>
        </div>

        <div className="cover-grid grid mt-3 mb-2 ">
          <div className="cover-grid-img">
            <img data-testid="Image" src={blogdetail.image.src} alt={blogdetail.image.alt} className="object-fit-cover" />
          </div>
        </div>

        <div>

          <p data-testid='discription' className="text-lg text-black mt-3 px-1 discriptiontext mb-5">
            {blogdetail.content}
          </p>
        </div>
      </div>

    </section>
  );
}
