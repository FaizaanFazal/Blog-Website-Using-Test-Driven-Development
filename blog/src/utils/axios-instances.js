// import axios from 'axios';

// export const axiosInstance = axios.create({
//   baseURL: 'http://localhost:8000/',
// });
import axios from 'axios';

const placeholderApi = axios.create({
  baseURL: 'http://localhost:8000',
});

export default placeholderApi;
