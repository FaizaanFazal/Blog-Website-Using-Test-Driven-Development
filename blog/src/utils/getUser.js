export const getUser = () => {
    const userString = localStorage.getItem('user');
    if (userString) {
      const user = JSON.parse(userString);
      return user;
    } else {
      return null;
    }
  };