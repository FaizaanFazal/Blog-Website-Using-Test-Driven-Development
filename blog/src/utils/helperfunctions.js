import { fireEvent } from '@testing-library/react';
import images from './images';

export const isProperImageURL = (url) => {
  /* eslint-disable-next-line no-useless-escape */
  const regex = /^(https?|ftp):\/\/(([a-z\d]([a-z\d-]*[a-z\d])?\.)+[a-z]{2,}|localhost)(\/[-a-z\d%_.~+]*)*(\?[;&a-z\d%_.~+=-]*)?(\#[-a-z\d_]*)?$/i;
  return regex.test(url);
};
export const isProperEmail = (email) => {
  const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
};
export const isProperPass = (pass) => {
  const passregex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  return passregex.test(pass);
};
export function backspace(element) {
  let actuallyTyped = element.value;

  const backspaceKey = {
    key: 'Backspace',
    code: 8,
    inputType: 'deleteContentBackward',
  };

  const sharedEventConfig = {
    key: backspaceKey.key,
    charCode: backspaceKey.code,
    keyCode: backspaceKey.code,
    which: backspaceKey.code,
    modifier: backspaceKey.modifier,
  };
  const downEvent = fireEvent.keyDown(element, sharedEventConfig);

  if (downEvent) {
    actuallyTyped = actuallyTyped.slice(0, -1);

    fireEvent.input(element, {
      target: { value: actuallyTyped },
      inputType: backspaceKey.inputType,
      bubbles: true,
      cancelable: true,
    });
  }

  fireEvent.keyUp(element, sharedEventConfig);
}

export const small = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: 'Elena Rodriguez' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    verified: true,
    slug: 'post1',
    content: "test Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.",
  },
];
export const greater = [
  {
    id: 1,
    title: '10 Tips for Budget-Friendly Travel',
    author: 'Elena Rodriguez',
    authorImage: { src: images.author1, alt: ' Elena' },
    date: '15 August',
    image: { src: images.blogImg1, alt: '10 Tips for Budget-Friendly Travel Blog Header Image' },
    slug: 'post1',
    verified: true,
    content: "Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n \n Traveling on a budget doesnt have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank.Traveling on a budget doesn't have to compromise your experience. Here are 10 tips to make the most of your journey without breaking the bank. \n",
  }];
