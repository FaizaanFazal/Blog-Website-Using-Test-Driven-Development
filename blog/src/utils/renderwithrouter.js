import {render} from '@testing-library/react'
import * as userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'

export const renderWithRouter = (ui, {route = '/'} = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    user: userEvent.setup,
    ...render(ui, {wrapper: MemoryRouter}),
  }
}