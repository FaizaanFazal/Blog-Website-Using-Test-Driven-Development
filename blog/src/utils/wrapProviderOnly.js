import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

const AllTheProviders = ({ children }) => (
  <Provider store={reduxStore}>
    {children}
  </Provider>
);

export const renderWithProvidersOnly = (ui, options) => render(<AllTheProviders>{ui}</AllTheProviders>, options);

// export function renderWithProviders(
//   ui,
//   {
//     // preloadedState = {},
//     // Automatically create a store instance if no store was passed in
//     store = reduxStore,
//     ...renderOptions
//   } = {},
// ) {
//   function Wrapper({ children }) {
//     return <Provider store={store}>{children}</Provider>;
//   }

//   // Return an object with the store and all of RTL's query functions
//   return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
// }

renderWithProvidersOnly.propTypes = {
  children: PropTypes.node.isRequired,
};
