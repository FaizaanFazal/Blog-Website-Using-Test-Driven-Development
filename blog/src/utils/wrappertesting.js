import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import reduxStore from '../Redux/store';
import '@testing-library/jest-dom';

export * from '@testing-library/react';

export function renderWithProviders(
  ui,
  {
    // preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = reduxStore,
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

renderWithProviders.propTypes = {
  children: PropTypes.node.isRequired,
};
