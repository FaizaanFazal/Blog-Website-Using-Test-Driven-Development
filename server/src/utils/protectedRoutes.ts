import { authenticateToken } from "@/models/authModel";
import { Request, Response, NextFunction, RequestHandler } from "express";


export const protectedRoutes = [
    {
        method: "",
        path: ""
    },
    {
        method: "",
        path: ""
    }];


//check protected routes to check authentication
export const isProtectedRouteMiddleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
    console.log(req.method)
    const protectedPathData = protectedRoutes.find((route) => route.path === req.path);
    if (protectedPathData) {
        const authenticatedUser = authenticateToken(req, res, next);
        if (authenticatedUser !== 403 && authenticatedUser !== 401) {
            req.user = authenticatedUser;
            console.log("User", authenticatedUser)
        } else {
            res.status(authenticatedUser);
            res.end();
        }
    } else {
        next();
    }
};