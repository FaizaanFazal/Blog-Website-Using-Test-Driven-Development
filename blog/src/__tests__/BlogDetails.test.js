import React from 'react';
import { screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { renderWithProviders } from '../utils/wrappertesting';
import BlogDetails from '../components/BlogDetails/BlogDetails';

describe('Unit tests for BlogDetails Components', () => {
  it('Snapshot test for BlogDetails component', () => {
    const { asFragment } = renderWithProviders(<BlogDetails />, { route: '/blogdetails/post1' });
    expect(asFragment()).toMatchSnapshot();
  });

  it('Dom test for rendering BlogDetails', () => {
    renderWithProviders(<BlogDetails />, { route: '/blogdetails/post1' });
    const detailseElement = screen.getByTestId('details');
    expect(detailseElement).toBeInTheDocument();
  });

  it('Image renders in BlogDetails component', () => {
    const history = createMemoryHistory();
    history.push('/post1');
    renderWithProviders(<BlogDetails />);
    const image = screen.getByTestId('Image');
    expect(image).toBeInTheDocument();
  });

  it('Image have ALT', () => {
    // const history = createMemoryHistory();
    // history.push('post1');
    // console.log(prettyDOM()); // Can be used to check what is being rendered
    renderWithProviders(<BlogDetails />, { route: '/blogdetails/post1' });
    const img = screen.getByTestId('Image');
    expect(img).toHaveAttribute('alt');
  });

  // it('Correct title is rendered as Heading', () => {
  //   const history = createMemoryHistory({
  //     initialEntries: ['/blogdetails/post1'],
  //     state: { slug: 'post1' },
  //   });
  //   /// / trying Router+provider
  //   renderWithBoth(<BlogDetails history={history} />);
  //   const titleheading = screen.getByTestId('titleHeading');
  //   expect(titleheading.textContent).toBe(greater[0].title);
  // });

  // it('Discription is rendered', () => {
  //   const history = createMemoryHistory();
  //   history.push({ slug: 'post1' });
  //   renderWithProviders(<BlogDetails history={history} />);
  //   const discription = screen.getByTestId('discription');
  //   expect(discription.textContent).toBe(greater[0].content);
  // });
});
