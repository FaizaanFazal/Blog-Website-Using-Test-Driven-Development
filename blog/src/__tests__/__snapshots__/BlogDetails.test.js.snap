// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for BlogDetails Components Snapshot test for BlogDetails component 1`] = `
<DocumentFragment>
  <section
    class="mb-5"
    data-testid="details"
  >
    <div
      class="container"
    >
      <h4
        class="title title-lg"
        data-testid="titleHeading"
      />
      <div
        class="card-footer  flex justify-between items-center"
      >
        <div
          class="writer-info grid"
        >
          <div
            class="info-avatar"
          >
            <img
              class="object-fit-cover"
            />
          </div>
          <div
            class="info-intro"
          >
            <p
              class="intro-name text-base"
            />
            <div
              class="intro-verify flex items-center"
            >
              <img
                alt="icon"
                class="verify-icon"
                src="verify_icon_light.svg"
              />
              <span
                class="verify-status text-sm text"
              >
                Not-Verified writer
              </span>
            </div>
          </div>
        </div>
        <div
          class="date-info text text-base"
          data-testid="detailsdate"
        >
          NA
        </div>
      </div>
      <div
        class="cover-grid grid mt-3 mb-2 "
      >
        <div
          class="cover-grid-img"
        >
          <img
            alt="none"
            class="object-fit-cover"
            data-testid="Image"
          />
        </div>
      </div>
      <div>
        <p
          class="text-lg text-black mt-3 px-1 discriptiontext mb-5 new-line"
          data-testid="discription"
        />
      </div>
    </div>
    <br />
  </section>
</DocumentFragment>
`;
